stdlib/public/core/Result.swift

```diff
@@ -0,0 +1,180 @@
+    case let .failure(error): return .failure(transform(error))
+    }
+  }
+  
+  public func flatMap<NewValue>(
+    _ transform: (Value) -> Result<NewValue, Error>
+  ) -> Result<NewValue, Error> {
+    switch self {
+    case let .success(value): return transform(value)
+    case let .failure(error): return .failure(error)
+    }
+  }
+  
+  public func flatMapError<NewError>(
+    _ transform: (Error) -> Result<Value, NewError>
+  ) -> Result<Value, NewError> {
+    switch self {
+    case let .success(value): return .success(value)
+    case let .failure(error): return transform(error)
+    }
+  }
+  
+  public func fold<Output>(
+    onSuccess: (Value) -> Output,
+    onFailure: (Error) -> Output
+    ) -> Output {
+    switch self {
+    case let .success(value): return onSuccess(value)
+    case let .failure(error): return onFailure(error)
+    }
+  }
+}
+
+extension Result where Error: Swift.Error {
+  public func unwrapped() throws -> Value {
+    switch self {
+    case let .success(value): return value
+    case let .failure(error): throw error
+    }
+  }
+}
+
+extension Result where Error == Swift.Error {
+  @_transparent
+  public init(_ throwing: () throws -> Value) {
+    do {
+      let value = try throwing()
+      self = .success(value)
+    } catch {
+      self = .failure(error)
+    }
+  }
+  
+  public func unwrapped() throws -> Value {
+    switch self {
+    case let .success(value): return value
+    case let .failure(error): throw error
+    }
+  }
+  
+  public func map<NewValue>(
+    _ transform: (Value) throws -> NewValue
+  ) -> Result<NewValue, Error> {
+    do {
+      switch self {
+      case let .success(value): return .success(try transform(value))
+      case let .failure(error): return .failure(error)
+      }
+    } catch {
+      return .failure(error)
+    }
+  }
+  
+  public func mapError(
+    _ transform: (Error) throws -> Swift.Error
+  ) -> Result<Value, Swift.Error> {
+    do {
+      switch self {
+      case let .success(value): return .success(value)
+      case let .failure(error): return .failure(try transform(error))
+      }
+    } catch {
+      return .failure(error)
+    }
+  }
+  
+  public func flatMap<NewValue>(
+    _ transform: (Value) throws -> Result<NewValue, Error>
+  ) -> Result<NewValue, Error> {
+    do {
+      switch self {
+      case let .success(value): return try transform(value)
+      case let .failure(error): return .failure(error)
+      }
+    } catch {
+      return .failure(error)
+    }
+  }
+  
+  public func flatMapError(
+    _ transform: (Error) throws -> Result<Value, Swift.Error>
+  ) -> Result<Value, Swift.Error> {
+    do {
+      switch self {
+      case let .success(value): return .success(value)
+      case let .failure(error): return try transform(error)
+      }
+    } catch {
+      return .failure(error)
+    }
+  }
+}
+
+extension Result: Equatable where Value: Equatable, Error: Equatable {
```

Comment:
I think you can just declare adoption of `Equatable` and `Hashable` without the implementations here.